Sources/SourceKitLSP/Swift/CodeCompletionSession.swift

```diff
@@ -25,10 +25,128 @@ import SourceKitD
 ///
 /// At the sourcekitd level, this uses `codecomplete.open`, `codecomplete.update` and
 /// `codecomplete.close` requests.
-actor CodeCompletionSession {
+class CodeCompletionSession {
+  // MARK: - Public static API
+
+  /// The queue on which all code completion requests are executed.
+  ///
+  /// This is needed because sourcekitd has a single, global code completion
+  /// session and we need to make sure that multiple code completion requests
+  /// don't race each other.
+  ///
+  /// Technically, we would only need one queue for each sourcekitd and different
+  /// sourcektid could serve code completion requests simultaneously.
+  ///
+  /// But it's rare to open multiple sourcekitd instances simultaneously and
+  /// even rarer to interact with them at the same time, so we have a global
+  /// queue for now to simplify the implementation.
+  private static let completionQueue = AsyncQueue(.serial)
+
+  /// The code completion session for each sourcekitd instance.
+  ///
+  /// `sourcekitd` has a global code completion session, that's why we need to
+  /// have a global mapping from `sourcekitd` to its currently active code
+  /// completion session.
+  ///
+  /// Modification of code completion sessions should only happen on
+  /// `completionQueue`.
+  private static var completionSessions: [ObjectIdentifier: CodeCompletionSession] = [:]
```

Comment:
Can't say I love this global state, is there any reason not to store it on `SourceKitDImpl` instead? FWIW the abstraction with the `SourceKitD` protocol doesn't seem super compelling to me since it seems like the only other conforming type is `FakeSourceKitD`, and that's only used to test the registry, it doesn't implement any of the other requirements. Seems like we could turn `SourceKitDRegistry` it into a generic `LibraryRegistry<T>` type or something, and then delete the `SourceKitD` protocol.

Reply:
Removing the abstraction of `SourceKitD` an `SourceKitDImpl` is also something that’s on my mind. Expect a follow-up PR for that sometime in the future.

> Can't say I love this global state, is there any reason not to store it on `SourceKitDImpl` instead?

It’s a layering decision: `CodeCompletionSession` is tied  quite closely to the needs of LSP (it takes `DocumentSnapshot` and returns `CompletionList`, both of which are defined in `LanguageServerProtocol`) while the `SourceKitD` module is a fairly low-level wrapper around the C sourcekitd API. So, I think `CodeCompletionSession` doesn’t belong in the `SourceKitD` module and thus it can’t be a member of `SourceKitD` (unless you are suggesting that `SourceKitDImpl` has something like `codeCompletionSession: Any` and I assume that’s not what you are suggesting).

Reply:
Hmm yeah that's fair, it's unfortunate because it seems like completion sessions are a sourcekid-specific concept. If we wanted to sink `CodeCompletionSession` down to a lower layer, it does seem like it doesn't have to return a `CompletionList`, it could return the raw response and let the language server handle transforming it. And it also seems like it doesn't need to store a full `DocumentSnapshot`, it only needs the `DocumentURI` I think (though maybe there's a use I'm missing). In particular it looks like the use in `guard snapshot.version == self.snapshot.version` is now redundant since that should be true by construction.

Either that or we could have a higher level wrapper that encapsulates both a `SourceKitD` and any session state it needs.